core/modules/shrscan/Director.h

```diff
@@ -0,0 +1,315 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_QSERV_SSCAN_DIRECTOR_H
+#define LSST_QSERV_SSCAN_DIRECTOR_H
+
+// System headers
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace lsst {
+namespace qserv {
+namespace shrscan {
+
+//-----------------------------------------------------------------------------
+//! @brief Describe a required table for a shared scan query.
+//-----------------------------------------------------------------------------
+
+class TableInfo
+{
+public:
+
+std::string tableName;    //< Name of the table
+bool        lockIndex;    //< If true, the table's index is locked, if any.
+bool        lockData;     //< If true, the table's data  is locked
+
+//-----------------------------------------------------------------------------
+//! Constructor
+//!
+//! @param  tabName   is the name of the table.
+//! @param  lkIndex   whether or not to lock the table's index
+//! @param  lkData    whether or not to lock the table's data
+//-----------------------------------------------------------------------------
+
+            TableInfo(const char *tabName="", bool lkIndex=true, bool lkData=true)
+                     : tableName(tabName), lockIndex(lkIndex), lockData(lkData)
+                     {}
+
+//-----------------------------------------------------------------------------
+//! Destructor
+//-----------------------------------------------------------------------------
+
+           ~TableInfo() {}
+};
+
+//-----------------------------------------------------------------------------
+//! @brief The Query pure abstract class describes a shared scan query.
+//-----------------------------------------------------------------------------
+
+class Query
+{
+public:
+
+//-----------------------------------------------------------------------------
+//! @brief The callback method to indicate a query has finished its run.
+//!
+//! The endQuery() method is called when the query completes or is canceled.
+//! This method is always called on a separate thread.
+//!
+//! @param  badChunks is a smart pointer to a vector of unprocessed chunks. This
+//!                   includes the chunk that was being processed if Run()
+//!                   returned false. You must take ownership of the shared
+//!                   pointer if you want to persist the vector upon return.
+//! @return canceled  When true, the query was cancelled and badChunks will be a
+//!                   zero length vector. Otherwise, the query completed and
+//!                   badChunks does hold the unprocessed chunk numbers.
+//-----------------------------------------------------------------------------
+
+virtual void endQuery(std::shared_ptr<std::vector<int>> const& badChunks,
```

Comment:
@jgates108  Multi-threading argument actually works the opposite way - it is *much* better to avoid unnecessary sharing in multi-threaded programs. If you can copy stuff and make it thread-private you have fewer problems to worry about. Shared pointer does not solve data sharing issue, it is of course thread safe, but it does not protect data that it points to.